### PR TITLE
[WB-73] Added dao for handling querying and storage of npm deps

### PR DESCRIPTION
**Pull request by ctdio**

### CODE REVIEW

File: src/dao/dependency/npm.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ * @param { Object } knex - the knex instance
+ * @param { String } key - the key to look for
+ * @param { Field } field - the jsonb field to search
+ *
+ * NOTE: field needs to be "manually" injected into the template field because
+ * knex will wrap the value in quotes if handled injected normally (thus causing
+ * the query to be malformed
+ */
+function _jsonKeyExistsIn (knex, key, field) {
+  return knex.raw(`${field}::jsonb ->> ? is not null`, [ key ])
```
Comment (Druotic) [inline src/dao/dependency/npm.js]: Is this type cast really necessary? Would simply ```knex.raw(`${field} ->> ? is not null`, [ key ])``` work as well?

On a side note, it looks like knex's query building support around json is a bit lacking :(

**Review by Druotic — CHANGES_REQUESTED** [review]

File: src/dao/dependency/npm.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+function _formatDependencies (data) {
+  return {
+    dependencies: JSON.stringify(data.dependencies || {}),
```
Comment (austinkelleher) [inline src/dao/dependency/npm.js]: Should we just store a null (not sure what the right term is in postgres land) if these are empty? We are going to have a lot of fields that simply store `{}` and never a actually store anything else.

**Review by austinkelleher — CHANGES_REQUESTED** [review]

File: src/models/dependencies/Npm.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/**
+* Model for Installations stored in the databases
```
[the remark this answers is not in the document]
Comment (charlieduong94) [reply src/models/dependencies/Npm.js]: whoops will fix

File: src/dao/migrations/20171114125010_create-npm-dependencies-table.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+const logger = require('~/src/logging').logger(module)
+const NpmProject = require('~/src/models/dependencies/Npm')
+
+const REPOSITORY_TABLE_NAME = NpmProject.typeName
+
+exports.up = function ({ schema }) {
+  logger.info('Attempting to run "up" on "create-repository-table"')
+  logger.info('Attempting to create repository table with table name: ', REPOSITORY_TABLE_NAME)
```
[the remark this answers is not in the document]
Comment (charlieduong94) [reply src/dao/migrations/20171114125010_create-npm-dependencies-table.js]: will fix

File: src/dao/dependency/npm.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+exports.findByDependency = async function (dependencyName) {
+  const knex = dao.getKnex()
+  const keyExistsIn = _jsonKeyExistsIn.bind(null, knex, dependencyName)
+
+  const results = await dao
+    .select('*')
+    .where(keyExistsIn('dependencies'))
+    .orWhere(keyExistsIn('dev_dependencies'))
+    .orWhere(keyExistsIn('peer_dependencies'))
+    .orWhere(keyExistsIn('optional_dependencies'))
```
[the remark this answers is not in the document]
Comment (charlieduong94) [reply src/dao/dependency/npm.js]: Yeah, do have some worries about this as well, but I feel like we should optimize this later once we actually reach that sort of scale. This _should_ still be much faster than performing something like `couchdb`'s design functions.

File: src/dao/dependency/npm.js (hunk 2 above)
Comment (charlieduong94) [reply src/dao/dependency/npm.js]: Storing `null` or `undefined` is fine. Will change.

File: src/dao/dependency/npm.js (hunk 1 above)
Comment (charlieduong94) [reply src/dao/dependency/npm.js]: Nope not needed, sorry thought that I needed to be explicit

**Review by austinkelleher — APPROVED** [review]

**Review by Druotic — APPROVED** [review]

**Pull request merged**